File: code/cpp/ch06/practice/practice_02_max_subarray_brute.cpp
/*
 * Practice 2: Max Subarray Sum (Brute Force)
 * ==========================================
 * Chapter 6: How Fast Is Your Code?
 *
 * PROBLEM:
 *   Given a vector of integers, find the contiguous subarray with the
 *   largest sum and return that sum.  Use the O(n^2) brute force
 *   approach (check all subarrays).  Return 0 for an empty array.
 *
 * EXAMPLES:
 *   solve({-2, 1, -3, 4, -1, 2, 1, -5, 4}) -> 6   (subarray [4,-1,2,1])
 *   solve({1})                              -> 1
 *   solve({-1, -2, -3})                     -> -1
 *   solve({})                               -> 0
 *
 * CONSTRAINTS:
 *   - 0 <= nums.size() <= 10^4
 *   - -10^4 <= nums[i] <= 10^4
 */

#include <iostream>
#include <vector>
using namespace std;

/**
 * Returns the maximum subarray sum using O(n^2) brute force.
 */
int solve(vector<int>& nums) {
    // TODO: Replace this with your solution
    return 0;
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
    cout << solve(nums) << endl;
    return 0;
}

File: code/java/ch06/practice/Practice02MaxSubarrayBrute.java
package ch06.practice;

import java.util.*;

/**
 * Practice 02: Max Subarray Sum (Brute Force)
 * =============================================
 * Chapter 6: How Fast Is Your Code?
 *
 * PROBLEM
 * -------
 * Given an integer array nums, find the contiguous subarray (containing
 * at least one number) which has the largest sum and return that sum.
 * Use the brute-force O(n^2) approach: try every possible start and end.
 *
 * For an empty array, return 0.
 *
 * INPUT FORMAT
 * ------------
 * A single line of space-separated integers (may be empty).
 *
 * OUTPUT FORMAT
 * -------------
 * Print the maximum subarray sum.
 *
 * CONSTRAINTS
 * -----------
 * 0 <= nums.length <= 10^4
 * -10^4 <= nums[i] <= 10^4
 *
 * EXAMPLES
 * --------
 * Input:  -2 1 -3 4 -1 2 1 -5 4    Output: 6   (subarray [4,-1,2,1])
 * Input:  1                          Output: 1
 * Input:  -1 -2 -3                   Output: -1
 * Input:  5 4 -1 7 8                 Output: 23
 *
 * HINT
 * ----
 * For each starting index i, compute the running sum as you extend to
 * each ending index j. Track the maximum sum seen.
 *
 * INSTRUCTIONS
 * ------------
 * Replace the "return 0;" in solve() with your solution.
 * The main method handles input/output -- don't change it.
 */
public class Practice02MaxSubarrayBrute {

    /**
     * Find the maximum subarray sum using brute force.
     *
     * @param nums the input array
     * @return maximum subarray sum, or 0 for empty array
     */
    public static int solve(int[] nums) {
        // TODO: Replace this with your solution
        return 0;
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String line = sc.nextLine().trim();
        if (line.isEmpty()) {
            System.out.println(solve(new int[0]));
        } else {
            int[] nums = Arrays.stream(line.split("\\s+"))
                               .mapToInt(Integer::parseInt).toArray();
            System.out.println(solve(nums));
        }
        sc.close();
    }
}
